pkg/server/credtemplate/builder_test.go

```diff
@@ -0,0 +1,1188 @@
+		overrideExpected func(expected *x509.Certificate)
+		expectErr        string
+	}{
+		{
+			desc: "defaults",
+		},
+		{
+			desc: "fail to get serial number",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.NewSerialNumber = failNewSerialNumber
+			},
+			expectErr: "failed to get new serial number: oh no",
+		},
+		{
+			desc: "invalid parent chain",
+			overrideParams: func(params *credtemplate.WorkloadX509SVIDParams) {
+				params.ParentChain = nil
+			},
+			expectErr: "parent chain required to build X509-SVID template",
+		},
+		{
+			desc: "empty SPIFFE ID",
+			overrideParams: func(params *credtemplate.WorkloadX509SVIDParams) {
+				params.SPIFFEID = spiffeid.ID{}
+			},
+			expectErr: "invalid X509-SVID ID: cannot be empty",
+		},
+		{
+			desc: "SPIFFE ID from another trust domain",
+			overrideParams: func(params *credtemplate.WorkloadX509SVIDParams) {
+				params.SPIFFEID = spiffeid.RequireFromString("spiffe://otherdomain.test/spire/agent/foo/foo-1")
+			},
+			expectErr: `invalid X509-SVID ID: "spiffe://otherdomain.test/spire/agent/foo/foo-1" is not a member of trust domain "domain.test"`,
+		},
+		{
+			desc: "invalid public key",
+			overrideParams: func(params *credtemplate.WorkloadX509SVIDParams) {
+				params.PublicKey = nil
+			},
+			expectErr: "x509: unsupported public key type: <nil>",
+		},
+		{
+			desc: "invalid DNS names",
+			overrideParams: func(params *credtemplate.WorkloadX509SVIDParams) {
+				params.PublicKey = nil
+			},
+			expectErr: "x509: unsupported public key type: <nil>",
+		},
+		{
+			desc: "override X509SVIDTTL",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.X509SVIDTTL = credtemplate.DefaultX509SVIDTTL * 2
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.NotAfter = now.Add(credtemplate.DefaultX509SVIDTTL * 2)
+			},
+		},
+		{
+			desc: "ttl capped by parent chain",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.X509SVIDTTL = parentTTL + time.Hour
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.NotAfter = now.Add(parentTTL)
+			},
+		},
+		{
+			desc: "override X509SVIDSubject",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.X509SVIDSubject = pkix.Name{CommonName: "OVERRIDE"}
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.Subject = pkix.Name{
+					CommonName: "OVERRIDE",
+					ExtraNames: []pkix.AttributeTypeAndValue{x509svid.UniqueIDAttribute(workloadID)},
+				}
+			},
+		},
+		{
+			desc: "with DNS names",
+			overrideParams: func(params *credtemplate.WorkloadX509SVIDParams) {
+				params.DNSNames = []string{"DNSNAME1", "DNSNAME2"}
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.DNSNames = []string{"DNSNAME1", "DNSNAME2"}
+				// CommonName is set to first DNS name by default
+				expected.Subject.CommonName = "DNSNAME1"
+			},
+		},
+		{
+			desc: "with DNS names and subject",
```

Comment:
maybe an overkill... 
but how about a test case with DNS and CommonName override? to be sure plugin wins